### ISSUE TITLE
onDelete and sync() depends on order of associations

### ISSUE THREAD

**Issue by seally1186** [sequelize/sequelize#6586]
This seems like it may be related to some issues already reported, but I couldn't quite find it. Apologies if it's a duplicate.
## What you are doing?

Here are two minimal models, and two orders of adding associations between them.

`````` js
const Sequelize = require("sequelize");
const sequelize = new Sequelize(/*...*/);

let a = sequelize.define("a", {
  id: { type: Sequelize.INTEGER, allowNull: false, primaryKey: true, autoIncrement: true },
}, {freezeTableName: true, timestamps: false});
let b = sequelize.define("b", {
  id: { type: Sequelize.INTEGER, allowNull: false, primaryKey: true, autoIncrement: true },
  aId: { type: Sequelize.INTEGER, allowNull: false }
}, {freezeTableName: true, timestamps: false});

if (process.argv[2] == "1") {
  a.hasMany(b);
  b.belongsTo(a);
} else {
  b.belongsTo(a);
  a.hasMany(b);
}
sequelize.sync({force: true, logging: console.log}).then(() => process.exit(0));```
``````
## What do you expect to happen?

The created foreign key probably shouldn't depend on whether we're in case 1 or 2.
## What is actually happening?

``` sql
$ node test.js 1 | grep CREATE
Executing (default): CREATE TABLE IF NOT EXISTS "a" ("id"   SERIAL , PRIMARY KEY ("id"));
Executing (default): CREATE TABLE IF NOT EXISTS "b" ("id"   SERIAL , "aId" INTEGER NOT NULL 
REFERENCES "a" ("id") ON DELETE CASCADE ON UPDATE CASCADE, PRIMARY KEY ("id"));

$ node test.js 2 | grep CREATE
Executing (default): CREATE TABLE IF NOT EXISTS "a" ("id"   SERIAL , PRIMARY KEY ("id"));
Executing (default): CREATE TABLE IF NOT EXISTS "b" ("id"   SERIAL , "aId" INTEGER NOT NULL 
REFERENCES "a" ("id") ON DELETE NO ACTION ON UPDATE CASCADE, PRIMARY KEY ("id"));
```

Note that the foreign key has `ON DELETE CASCADE` in case 1, but `ON DELETE NO ACTION` in case 2. I think the problem is that `hasMany` (and `hasOne`?) default to `onDelete: "CASCADE"`, while `belongsTo` (and `belongsToMany`?) default to `onDelete: "NO ACTION"`.

**Dialect:** postgres
**Database version:** 9.3.14
**Sequelize version:** 3.24.2

**Comment by felixfbecker** [comment]
I agree this is confusing, but it can be circumvented by always being explicit about what you want.

**Comment by seally1186** [comment]
Thanks for the reply.

Yeah, I agree I can always set the onDelete on every association. But I guess I'm not seeing the intention behind having belongsTo and hasMany use different defaults for that property? Is it being used somewhere other than generating the foreign key constraint? Isn't the most common use case going to be defining both directions of the relation?

**Comment by dsbert** [comment]
According to the documentation, the default for any delete operation is either `SET NULL` or `CASCADE`. Therefore, defaulting to `NO ACTION` seems like a bug (or the docs need updating).

> For 1:1 and 1:m associations the default option is SET NULL for deletion, and CASCADE for updates. For n:m, the default for both is CASCADE. This means, that if you delete or update a row from one side of an n:m association, all the rows in the join table referencing that row will also be deleted or updated.

**Comment by dsbert** [comment]
I do not know if this was ever fixed because I no longer use sequelize.

**Comment by ajones513** [comment]
I've hit this too, my reproduction case FWIW:

```js
const Sequelize = require('sequelize');
const sequelize = new Sequelize('db', 'user', 'password', {
  host: 'localhost',
  dialect: 'mysql',
  operatorsAliases: false
});

const a = sequelize.define('a');
const b = sequelize.define('b');
const c = sequelize.define('c');

// Gives a.bId 'ON DELETE CASCADE'
b.hasMany(a, { foreignKey: { allowNull: false } });
a.belongsTo(b, { foreignKey: { allowNull: false } });

// Gives c.bId 'ON DELETE NO ACTION'
c.belongsTo(b, { foreignKey: { allowNull: false } });
b.hasMany(c, { foreignKey: { allowNull: false } });

sequelize.sync({ force: true });
```

**Dialect**: mysql
**Database version**: 5.7.21
**Sequelize version**: 4.34.0

**Comment by sant123** [comment]
Any advance on this? Just asking 😝

**Comment by sant123** [comment]
Shall this be added to the next milestone? https://github.com/sequelize/sequelize/milestone/18

**Comment by ryan-harris** [comment]
I believe the problem is that belongTo's default value of ON DELETE in the code is SET NULL or NO ACTION.  The documentation clearly states that for 1:1 and 1:m associations the default value for delete should be SET NULL or CASCADE (depending if the foreign key allows null).  The other 1:1 and 1:m associations (hasOne, hasMany) have the correct default value for ON DELETE.


The proper solution might also involve not setting an association's options to the defaults until sync() is called, so that with a pair of associations, the 2nd association evaluated can overwrite those options if they weren't explicitly defined by the user on the 1st association. That way I don't have to define all the same options on both associations of a pair, not knowing which model/association would get evaluated first (using 'sequelize init:models' the index.js loads the models and their associations in an ordering that isn't guaranteed, so right now I have to defined the options on both Models to be safe)
